### PR TITLE
Fix Yarn container log and livy submission log fetcher issues

### PR DESCRIPTION
**Pull request by wezhang**

### CODE REVIEW

**Comment by wezhang** [comment]
```
java.lang.RuntimeException: /home/travis/.azure/azureProfile.json (No such file or directory)
	at com.microsoft.azure.spark.tools.http.AzureOAuthTokenFetcher.buildFromAzureCli(AzureOAuthTokenFetcher.java:41)
```

File: src/test/resources/com/microsoft/azure/spark/tools/job/LivySparkBatchScenario.feature [hunk 1]
```diff
@@ -19,3 +19,50 @@ Feature: LivySparkBatch unit tests
...
+  Scenario: await Spark job is done behavior
+    Given setup a mock Livy service with the following scenario 'awaitJobIsDoneUT'
+      | ACTION | URI      | RESPONSE_STATUS | RESPONSE_BODY                     | PREV_STATE | NEXT_STATE |
+      | GET    | /batch/9 | 200             | {"id": 9, "state": "not_started"} | Started    | starting_1 |
+      | GET    | /batch/9 | 200             | {"id": 9, "state": "starting"}    | starting_1 | starting_2 |
+      | GET    | /batch/9 | 200             | {"id": 9, "state": "starting"}    | starting_2 | starting_3 |
+      | GET    | /batch/9 | 200             | {"id": 9, "state": "running"}     | starting_3 | running_1  |
+      | GET    | /batch/9 | 200             | {"id": 9, "state": "running"}     | running_1  | running_2  |
+      | GET    | /batch/9 | 200             | {"id": 9, "state": "success"}     | running_2  | success_1  |
+      | GET    | /batch/9 | 200             | {"id": 9, "state": "dead"}        | success_1  | end        |
```
Comment (t-rufang) [inline src/test/resources/com/microsoft/azure/spark/tools/job/LivySparkBatchScenario.feature]: I suppose the job state shouldn't switch to `dead` state if it's already in `success` state.

File: src/test/resources/com/microsoft/azure/spark/tools/job/LivySparkBatchScenario.feature [hunk 2]
```diff
@@ -19,3 +19,50 @@ Feature: LivySparkBatch unit tests
...
+      | \\nstderr: |
+      | \\nYARN Diagnostics: |
+    Then check parsed Livy logs stdout should be
+      | stdout:  |
+      | one      |
+      | two      |
+    Then check parsed Livy logs stderr should be
+      | \\nstderr: |
+    Then check parsed Livy logs yarn diagnostics should be
+      | \\nYARN Diagnostics: |
```
Comment (t-rufang) [inline src/test/resources/com/microsoft/azure/spark/tools/job/LivySparkBatchScenario.feature]: Add a line break here would be better

File: src/test/resources/recordings/com.microsoft.azure.spark.tools.processes.SparkBatchJobRemoteProcessScenario/mappings/livy_batches_261-0a40e506-d70c-4faf-89c4-c52eeff08582.json [hunk 3]
```diff
@@ -1,8 +1,8 @@
 {
```
Comment (t-rufang) [inline src/test/resources/recordings/com.microsoft.azure.spark.tools.processes.SparkBatchJobRemoteProcessScenario/mappings/livy_batches_261-0a40e506-d70c-4faf-89c4-c52eeff08582.json]: Do we need to keep these recording json files?

File: src/main/java/com/microsoft/azure/spark/tools/job/LivySparkBatch.java [hunk 4]
```diff
@@ -373,20 +425,23 @@ public URI getUri() {
...
         return Observable.fromCallable(this::getUri)
                 .flatMap(uri -> getHttp()
-                        .delete(uri.toString(), null, getHeadersToAddOrReplace()));
+                        .delete(uri.toString(), emptyList(), getHeadersToAddOrReplace()));
     }
 
-    private Observable<Batch> getSparkBatchRequest() {
+    private Observable<Batch> getSparkBatchRequest(final String caller) {
+        final List<Header> headers = new ArrayList<>(getHeadersToAddOrReplace());
+        headers.add(new BasicHeader("X-Invoked-By-Method", caller));
```
[the remark this answers is not in the document]
Comment (wezhang) [reply src/main/java/com/microsoft/azure/spark/tools/job/LivySparkBatch.java]: It's used for helping debug HTTP requests to know where it came from.

File: src/main/java/com/microsoft/azure/spark/tools/job/LivySparkBatch.java [hunk 5]
```diff
@@ -324,15 +373,14 @@ public boolean isSuccess(final String toCheck) {
...
-                    log().debug("Deploy " + 1 //getDelaySeconds()
-                            + " seconds for next job status probe");
-                    return ob.delay(
-                            1, //getDelaySeconds(),
-                            TimeUnit.SECONDS);
-                })
-                .takeUntil(batch -> !isDone(batch.state))
-                .filter(batch -> !isDone(batch.state))
+                .repeatWhen(ob -> ob
+                        .doOnNext(v -> log().debug("Deploy " + 1 //getDelaySeconds()
```
[the remark this answers is not in the document]
Comment (wezhang) [reply src/main/java/com/microsoft/azure/spark/tools/job/LivySparkBatch.java]: No changes here. We can fix it later with different delay settings.

File: src/main/java/com/microsoft/azure/spark/tools/job/LivySparkBatch.java [hunk 6]
```diff
@@ -400,11 +455,12 @@ public URI getUri() {
...
-        getLaterBatchId().setIfNull(batch.getId());
+        this.getLaterBatchId().setIfNull(batch.getId());
+        if (batch.getAppId() != null) {
+            this.getLaterAppId().setIfNull(batch.getAppId());
+        }
         this.state = batch.getState();
-        this.appId = batch.getAppId();
-        this.appInfo = batch.getAppInfo();
-        this.submissionLogs = (batch.getLog() == null) ? Collections.emptyList() : batch.getLog();
+        this.submissionLogs = (batch.getLog() == null) ? emptyList() : batch.getLog();
```
[the remark this answers is not in the document]
Comment (wezhang) [reply src/main/java/com/microsoft/azure/spark/tools/job/LivySparkBatch.java]: Good catch! Fix it.

File: src/main/java/com/microsoft/azure/spark/tools/job/YarnContainerLogFetcher.java [hunk 7]
```diff
@@ -148,112 +179,79 @@ private boolean isYarnAppLogAggregationDone(final App yarnApp) {
...
-                    }
-
-                    String probedLogUrl = this.getCurrentLogUrl();
-                    if (probedLogUrl == null) {
-                        return Observable.empty();
-                    }
+                .delay(logUrl -> getYarnApp())  // Get App state before fetching
+                .doOnNext(this::setCurrentLogUrl)
+                .flatMap(logUrl -> this.getContentFromYarnLogDom(logUrl, type, logOffset, size))
+                .repeatWhen(completed -> completed
```
[the remark this answers is not in the document]
Comment (wezhang) [reply src/main/java/com/microsoft/azure/spark/tools/job/YarnContainerLogFetcher.java]: `repeatWhen` and `first` to get the first non-empty event

File: src/main/java/com/microsoft/azure/spark/tools/processes/SparkBatchJobRemoteProcess.java [hunk 8]
```diff
@@ -97,7 +97,7 @@ public int exitValue() {
     @Override
     public void destroy() {
```
[the remark this answers is not in the document]
Comment (wezhang) [reply src/main/java/com/microsoft/azure/spark/tools/processes/SparkBatchJobRemoteProcess.java]: Fix it.

File: src/test/resources/com/microsoft/azure/spark/tools/job/LivySparkBatchScenario.feature (hunk 1 above)
Comment (wezhang) [reply src/test/resources/com/microsoft/azure/spark/tools/job/LivySparkBatchScenario.feature]: It's a disturbing case. The final state should arrive that.

File: src/test/resources/com/microsoft/azure/spark/tools/job/LivySparkBatchScenario.feature (hunk 2 above)
Comment (wezhang) [reply src/test/resources/com/microsoft/azure/spark/tools/job/LivySparkBatchScenario.feature]: Fix it with a new scenario.

File: src/test/resources/recordings/com.microsoft.azure.spark.tools.processes.SparkBatchJobRemoteProcessScenario/mappings/livy_batches_261-0a40e506-d70c-4faf-89c4-c52eeff08582.json (hunk 3 above)
Comment (wezhang) [reply src/test/resources/recordings/com.microsoft.azure.spark.tools.processes.SparkBatchJobRemoteProcessScenario/mappings/livy_batches_261-0a40e506-d70c-4faf-89c4-c52eeff08582.json]: They are used for replay in UT.

**Pull request merged**